src/functional/ffront/foast_passes/type_deduction.py

```diff
@@ -191,17 +191,21 @@ def visit_Symbol(
 
     def visit_Subscript(self, node: foast.Subscript, **kwargs) -> foast.Subscript:
         new_value = self.visit(node.value, **kwargs)
-        new_type = None
-        if kwargs.get("in_shift", False):
-            return foast.Subscript(
-                value=new_value,
-                index=node.index,
-                type=new_value.type,
-                location=node.location,
-            )
+        new_type: Optional[ct.SymbolType] = None
         match new_value.type:
             case ct.TupleType(types=types):
                 new_type = types[node.index]
+            case ct.OffsetType(source=source, target=(target1, target2)):
+                if not target2.local:
+                    raise FieldOperatorTypeDeductionError.from_foast_node(
+                        new_value, msg="Second dimension in offset must be a local dimension."
+                    )
+                new_type = ct.OffsetType(source=source, target=(target1,))
+            case ct.OffsetType(source=_, target=(_,)):
+                # for cartesian axes (e.g. I, J) the index of the subscript only
+                #  signifies the displacement in the respective dimension,
+                #  but does not change the target type.
+                new_type = new_value.type
```

Comment:
It seems there might be a loop hole here, where `FieldOffset(source=X, target=[Y])` would silently be treated as `FieldOffset(source=X, target=[X])`.

It looks like it should either be disallowed or treated differently. If it is disallowed and caught at an earlier time, then this is fine here.